Exercise-02/Exercise02/main.swift

```diff
@@ -22,18 +27,167 @@ func printInvalidCommand() {
+}
+
+class DayInfo {
+  private var timeToTemperature: [Date: Double] = [:]
+  
+  func addData(dateAndTime: Date, temperature: Double) throws {
+    guard timeToTemperature.count < 6 else {
+      throw TemperatureError.validationFailed("reached maximum number of measures per day")
+    }
+    
+    if let minDistance = (timeToTemperature.keys.map { (otherDateAndTime: Date) -> Double in abs(dateAndTime.timeIntervalSince(otherDateAndTime))}.min()) {
+      let hoursDistance = Int(minDistance / 3600)
+      guard hoursDistance >= 4 else {
+        throw TemperatureError.validationFailed("interval between measures should be at least 4 hours")
+      }
+    }
+    
+    timeToTemperature[dateAndTime] = temperature
+  }
+  
+  func removeData(dateAndTime: Date) throws -> Double {
+    guard let removedTemperature = timeToTemperature.removeValue(forKey: dateAndTime) else {
+      throw TemperatureError.noData("no data found for the given time")
+    }
+    
+    return removedTemperature
+  }
+  
+  func sum() -> Double {
+    var sum = 0.0
+    for temperature in timeToTemperature.values {
+      sum += temperature
+    }
+    
+    return sum
+  }
+  
+  func count() -> Int {
+    return timeToTemperature.count
+  }
+  
+  func minTemperature() -> Double {
+    var min = Double.infinity
+    for temp in timeToTemperature.values {
+      min = Double.minimum(min, temp)
+    }
+    return min
+  }
+  
+  func maxTemperature() -> Double {
+    var max = -Double.infinity
+    for temp in timeToTemperature.values {
+      max = Double.maximum(max, temp)
+    }
+    return max
+  }
+  
+  func sorted(ascending: Bool) -> [(key: Date, value: Double)] {
+    return timeToTemperature.sorted(by: { ascending ? $0.key < $1.key : $0.key > $1.key })
+  }
+}
+
+class CityInfo {
+  var minTemperature: Double?
+  var maxTemperature: Double?
+  var dateToDayInfo: [Date: DayInfo] = [:]
+  
+  func addData(dateAndTime: Date, temperature: Double) throws {
+    var dayInfo: DayInfo
+    if let dayInfoFromCityInfo = getDayInfo(dateAndTime: dateAndTime) {
+      dayInfo = dayInfoFromCityInfo
+    } else {
+      dayInfo = DayInfo()
+      dateToDayInfo[getCalendar().startOfDay(for: dateAndTime)] = dayInfo
+    }
+    
+    try dayInfo.addData(dateAndTime: dateAndTime, temperature: temperature)
+    minTemperature = Double.minimum(minTemperature ?? Double.infinity, temperature)
+    maxTemperature = Double.maximum(maxTemperature ?? -Double.infinity, temperature)
+  }
+  
+  func removeData(dateAndTime: Date) throws {
+    guard let dayInfo = getDayInfo(dateAndTime: dateAndTime) else {
+      throw TemperatureError.noData("no data found for the given day")
+    }
+    
+    let removedTemperature = try dayInfo.removeData(dateAndTime: dateAndTime)
+    
+    // Assuming temperature removal is not a common operation, we can afford to recompute minimum and/or maximum in case the current minimum or maximum is removed
+    
+    if let minValue = minTemperature {
```

Comment:
```suggestion
    if let minValue = self.minTemperature {
```
*always* use `self.`, in order to underline that is not a local variable.